File: core/src/main/java/com/whylabs/logging/core/statistics/NumberTracker.java
package com.whylabs.logging.core.statistics;

import com.google.protobuf.ByteString;
import com.whylabs.logging.core.format.NumbersMessage;
import com.whylabs.logging.core.statistics.datatypes.DoubleTracker;
import com.whylabs.logging.core.statistics.datatypes.LongTracker;
import com.whylabs.logging.core.statistics.datatypes.VarianceTracker;
import java.util.Optional;
import lombok.AccessLevel;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.val;
import org.apache.datasketches.memory.Memory;
import org.apache.datasketches.memory.WritableMemory;
import org.apache.datasketches.quantiles.DoublesSketch;
import org.apache.datasketches.quantiles.UpdateDoublesSketch;
import org.apache.datasketches.theta.UpdateSketch;

@Getter
@Builder(setterPrefix = "set")
@AllArgsConstructor(access = AccessLevel.PRIVATE)
public class NumberTracker {
  // our own trackers
  VarianceTracker variance;
  DoubleTracker doubles;
  LongTracker longs;

  // sketches
  UpdateDoublesSketch histogram; // histogram
  UpdateSketch thetaSketch;

  public NumberTracker() {
    this.variance = new VarianceTracker();
    this.doubles = new DoubleTracker();
    this.longs = new LongTracker();

    this.thetaSketch = UpdateSketch.builder().build();
    this.histogram = DoublesSketch.builder().setK(256).build();
  }

  public void track(Number number) {
    double dValue = number.doubleValue();
    variance.update(dValue);
    thetaSketch.update(dValue);
    histogram.update(dValue);

    if (doubles.getCount() > 0) {
      doubles.update(dValue);
    } else if (number instanceof Long || number instanceof Integer) {
      longs.update(number.longValue());
    } else {
      doubles.addLongs(longs);
      longs.reset();
      doubles.update(dValue);
    }
  }

  public NumbersMessage.Builder toProtobuf() {
    val builder =
        NumbersMessage.newBuilder()
            .setVariance(variance.toProtobuf())
            .setTheta(ByteString.copyFrom(thetaSketch.toByteArray()))
            .setHistogram(ByteString.copyFrom(histogram.toByteArray()));

    if (this.doubles.getCount() > 0) {
      builder.setDoubles(this.doubles.toProtobuf());
    } else if (this.longs.getCount() > 0) {
      builder.setLongs(this.longs.toProtobuf());
    }

    return builder;
  }

  public static NumberTracker fromProtobuf(NumbersMessage message) {
    val tMem = WritableMemory.wrap(message.getTheta().toByteArray());
    val hMem = Memory.wrap(message.getHistogram().toByteArray());
    val builder =
        NumberTracker.builder()
            .setThetaSketch(UpdateSketch.wrap(tMem))
            .setHistogram(UpdateDoublesSketch.heapify(hMem))
            .setVariance(VarianceTracker.fromProtobuf(message.getVariance()));

    Optional.ofNullable(message.getDoubles())
        .map(DoubleTracker::fromProtobuf)
        .ifPresent(builder::setDoubles);
    Optional.ofNullable(message.getLongs())
        .map(LongTracker::fromProtobuf)
        .ifPresent(builder::setLongs);

    return builder.build();
  }
}

File: whylogs/logging/core/statistics/numbertracker.py
#!/usr/bin/env python3
"""
created 5/5/20 by ibackus 
"""
import datasketches
from whylogs.logging.core.statistics.datatypes import VarianceTracker
from whylogs.logging.core.data import NumbersMessage


class NumberTracker:
    """

    """
    def __init__(self):
        # TODO: add keyword argument intialization
        # Our own trackers.  TODO: implement our own trackers
        self.variance = VarianceTracker()
        self.doubles = None
        self.longs = None

        # Sketches
        self.thetaSketch = datasketches.update_theta_sketch()
        self.histogram = None

    def track(self, number):
        self.variance.update(number)
        self.thetaSketch.update(number)
        # TODO: histogram update
        # TODO: update self.doubles and self.longs

    def toProtobuf(self):
        """
        """
        msg = NumbersMessage(
            variance=self.variance.toProtobuf(),
            theta=self.thetaSketch.serialize(),
            # TODO: add doubles, longs, and histogram
        )
        return msg

    @staticmethod
    def fromProtobuf(message):
        """
        """
        theta_sketch = datasketches.update_theta_sketch.deserialize(
            message.theta)
        tracker = NumberTracker()
        tracker.thetaSketch = theta_sketch
        tracker.variance = VarianceTracker.fromProtobuf(message.variance)
        # TODO: de-serialize histogram, and implement "doubles/longs" trackers

        return tracker


